### PR TITLE
Update cookiecutter to use GitHub Actions

### PR DESCRIPTION
**Pull request by trallard**
This PR relates to <https://github.com/jupyter/docker-stacks/issues/1139>.

I am working on swapping from Travis to GH Actions, however there are a couple of things I wanted to check before moving forward:

- The current version of the cookie-cutter uses Python 3.6. To keep in sync with `docker-stacks` I am upgrading to `3.8`

**For the actual `{{cookiecutter}}`**
-  I made this initial draft using the official `docker` action to login to Docker Hub.  I have been using this on other repos and has been working well so far.
- I noticed there is a [post_push hook](https://github.com/jupyter/cookiecutter-docker-stacks/blob/master/%7B%7Bcookiecutter.stack_name%7D%7D/hooks/post_push) which uses the sha to tag the image with the `SHA` and push to Docker. Though I do not think this currently in use within the Travis CI. I recently did some work on Docker for SciPy and do the build + push directly from within GH Actions (see https://github.com/scipy/scipy/blob/master/.github/workflows/docker.yml for reference) I could follow a similar approach here but do not want to impose any changes without discussing first

### CODE REVIEW

**Comment by parente** [comment]
Thanks for working on this @trallard. I totally forgot this cookiecutter was in a separate repo from jupyter/docker-stacks. (I've been living in monorepos lately.) I'll add the other maintainers of that docker-stacks here so pull requests like this one don't go unnoticed in my GitHub notifications alone for so long.

> The current version of the cookie-cutter uses Python 3.6. To keep in sync with docker-stacks I am upgrading to 3.8

+1 Feel free to move to the latest stable Python release if it's easy as well.

> I made this initial draft using the official docker action to login to Docker Hub. I have been using this on other repos and has been working well so far.

+1 Makes sense to use the official actions when available. Ideally this cookiecutter would give users the ability to push images to any registry, but that's definitely out of scope of this PR. (Noting only as a potential future project for anyone following along.)

> I noticed there is a post_push hook which uses the sha to tag the image with the SHA and push to Docker. Though I do not think this currently in use within the Travis CI. I recently did some work on Docker for SciPy and do the build + push directly from within GH Actions (see https://github.com/scipy/scipy/blob/master/.github/workflows/docker.yml for reference) I could follow a similar approach here but do not want to impose any changes without discussing first

The original CI/CD implementation in jupyter/docker-stacks built images on Travis for pull requests only, but then used Docke r Hub CI to build merges to the main branch and push to the Docker Hub registry. The [`post_push` hook is a script honored by builds done on Docker Hub](https://docs.docker.com/docker-hub/builds/advanced/#custom-build-phase-hooks) and was what we used to apply the git SHA tags to images build there. This entier design was a giant workaround for build time limits on Travis and is ripe for replacement.

I think it'd be best to move the cookiecutter to building and pushing PR and main branch images from GitHub actions like we now do in jupyter/docker-stacks. I think it's fine to use whatever technique you think works best to apply the git SHA as a tag given that we'd be replacing the entire CI/CD implementation anyway.

**Comment by trallard** [comment]
I have finally made time to finish this off.  I went ahead with the changes discussed above. However, there are some outstanding actions / decisions:
- I wrote the GH action so that it pushes a latest and a SHA tagged image
- With this action then the contents in `hooks` becomes irrelevant and can be removed
- The `travis.yml` files can now be deleted but I wanted to check if you all are all happy with this implementation before going ahead

**Review by romainx — CHANGES_REQUESTED** [review]
Hello @trallard,

Many thanks for this PR. I've done a couple of fixes / addition.

- `docker.yml`
  - Create two distinct steps build and push to be able to test in between.
  - Use `driver: docker` to be able to test locally.
  - Fix cookiecutter variables in `push` where not converted.
  - Fix line break introduced by `raw` blocks.
  - Add main branch along with master to trigger the build / push on the main branch.
- `Makefile`
  - Fix `dev` target
  - Add run target to run a shell (for debugging purpose)
- Add default `.dockerignore`

I've tested and successfully built an image from this template. 

I think you can now safely remove

- [ ] `hooks`:  they are useless and moreover this system will be replaced in docker stack. If users need to add additional tags, they will be able to add them in the GitHub `docker.yml` file.
- [ ] `.travis.yml` files (in template and main repo) since they will be replaced by corresponding GitHub actions.

After that I think we will be good for the merge. If we need to refine or improve, we will do it in a follow-on PR.

Thanks.
Best

File: .github/workflows/CI.yml
Comment (mathbunnyru) [inline .github/workflows/CI.yml:5]: Should we add `main` branch here as well?

File: {{cookiecutter.stack_name}}/.github/workflows/docker.yml
Comment (mathbunnyru) [inline {{cookiecutter.stack_name}}/.github/workflows/docker.yml:6]: Should we filter branch here as well?

File: .github/workflows/CI.yml
Comment (mathbunnyru) [inline .github/workflows/CI.yml:27]: Please add an empty line, so github doesn't complain

**Review by mathbunnyru — APPROVED** [review]
Good job @trallard and @romainx as well. I'm not fully available and equipped at the moment, but I wrote few minor things. Feel free to ignore them.

File: .github/workflows/CI.yml
Comment (romainx) [reply .github/workflows/CI.yml:5]: We can but this repo main branch is `master` so there is not impact.

File: {{cookiecutter.stack_name}}/.github/workflows/docker.yml
Comment (romainx) [reply {{cookiecutter.stack_name}}/.github/workflows/docker.yml:6]: I don't think so.

File: .github/workflows/CI.yml
Comment (romainx) [reply .github/workflows/CI.yml:27]: 👍

Comment (romainx) [reply .github/workflows/CI.yml]: Done

File: .github/workflows/CI.yml
Comment (romainx) [reply .github/workflows/CI.yml:5]: Done anyway 😄

**Review by mathbunnyru — APPROVED** [review]

**Comment by trallard** [comment]
Thanks for your reviews @romainx  and @mathbunnyru I went ahead and deleted the `travis.yaml` files and the `hooks`

Unless there is anything else outstanding this should be good to merge

**Review by mathbunnyru — APPROVED** [review]

**Review by romainx — APPROVED** [review]

**Comment by romainx** [comment]
Hello Ok let's merge it. We will fix issues -- if any -- in follow-on PR.
@trallard & @mathbunnyru thanks for this work.

**Pull request merged**
